lib/request.js

```diff
@@ -0,0 +1,102 @@
+'use strict'
+const http = require('http')
+const https = require('https')
+const http2 = require('http2')
+const eos = require('end-of-stream')
+const pump = require('pump')
+const { stripHttp1ConnectionHeaders } = require('./utils')
+
+class RequestAgent {
+  constructor (opts) {
+    this.http2 = !!opts.http2
+    this.requests = {
+      'http:': http,
+      'https:': https
+    }
+    if (this.http2) {
+      if (!opts.base) throw new Error('Option base is required when http2 is true')
+      this.http2Client = http2.connect(opts.base, { rejectUnauthorized: opts.rejectUnauthorized })
+    } else {
+      this.agents = {
+        'http:': new http.Agent(agentOption(opts)),
+        'https:': new https.Agent(agentOption(opts))
+      }
+    }
+  }
+
+  handleHttp1Req ({ method, url, qs, headers, body }, done) {
+    const req = this.requests[url.protocol].request({
+      method,
+      port: url.port,
+      path: url.pathname + qs,
+      hostname: url.hostname,
+      headers,
+      agent: this.agents[url.protocol]
+    })
+    req.on('error', done)
+    req.on('response', (res) => {
+      done(null, { statusCode: res.statusCode, headers: res.headers, stream: res })
+    })
+    end(req, body, done)
+  }
+
+  handleHttp2Req ({ method, url, qs, headers, body }, done) {
+    this.http2Client.ref()
+    const isGet = (method === 'GET' || method === 'get')
+    const req = this.http2Client.request({
+      ':method': method,
+      ':path': url.pathname + qs,
+      ...stripHttp1ConnectionHeaders(headers)
+    })
+    if (!isGet) {
+      end(req, body, done)
+    }
+    req.on('error', done)
+    eos(req, () => { this.http2Client.unref() })
```

Comment:
Why is `unref()` needed?

Reply:
https://github.com/mcollina/h2url/blob/c23d1e920308f68ef2ba5d49720adc54326dd1b5/index.js#L28
I went by the comment on above. If not needed I will remove it. May be I mis-interpreted.